theano/tensor/nnet/ctc.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import (division, absolute_import, print_function)
+import os
+import os.path
+import theano.tensor as T
+from theano import config
+from theano import gof
+from theano.gof import local_optimizer
+from theano.gof.cmodule import GCC_compiler
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+from theano.tensor.extra_ops import cpu_contiguous
+from theano.gradient import grad_undefined
+
+
+def _ctc_find_lib():
+    """
+    Find the directory that contains libwarpctc.so
+    """
+    if config.ctc.root != '':
+        for lib_dir in ["build", "lib", "lib64"]:
+            lib_path = os.path.join(config.ctc.root, lib_dir)
+            if os.path.isdir(lib_path) and os.path.exists(lib_path):
+                lib_found = os.path.exists(os.path.join(lib_path, "libwarpctc.so"))
+                if lib_found:
+                    return lib_path
+    return None
+
+
+def _ctc_check_compile(ctc_lib_path):
+    preambule = """
+#include <string.h>
+#include "ctc.h"
+"""
+
+    body = """
+ctcOptions options;
+memset(&options, 0, sizeof(ctcOptions));
+options.loc = CTC_CPU;
+options.num_threads = 1;
+"""
+
+    params = ['-I%s' % (os.path.dirname(__file__))]
+    if ctc_lib_path is not None:
+        params.extend(["-I%s" % (os.path.join(config.ctc.root, "include"))])
+        params.extend(["-L%s" % (ctc_lib_path)])
+    params.extend(["-l", "warpctc"])
+    compiler_res = GCC_compiler.try_flags(
+        params, preambule=preambule, body=body,
+        try_run=False, output=True)
+
+    avail, out, err = compiler_res if isinstance(compiler_res, tuple) else (compiler_res, None, None)
+    if not avail:
+        return False, ("cannot compile with warp-ctc. "
+                       "We got this error:\n" + str(err))
+    return True, None
+
+
+def ctc_present():
+    if ctc_present.avail is not None:
+        return ctc_present.avail
+    ctc_lib_path = _ctc_find_lib()
+    ctc_present.path = ctc_lib_path
+    ctc_present.avail, ctc_present.msg = _ctc_check_compile(ctc_present.path)
+    return ctc_present.avail
+
+
+ctc_present.avail = None
+ctc_present.msg = None
+ctc_present.path = None
+
+
+def ctc_available():
+    if os.name == 'nt':
+        ctc_available.msg = 'Windows platforms are currently not supported ',
+        'by underlying CTC library (warp-ctc).'
+        return False
+    elif not ctc_present():
```

Comment:
In fact, the ``ctc_available`` function calls the ``_ctc_check_compile`` function, which tries to compile a sample code using the CTC library.